hw/drivers/sensors/tsl2591/src/tsl2591.c

```diff
@@ -0,0 +1,706 @@
+            rc = SYS_EINVAL;
+            goto err;
+    }
+
+    rc = tsl2591_get_integration_time(itf, &int_time);
+    if (rc) {
+        goto err;
+    }
+
+    rc = tsl2591_write8(itf, TSL2591_COMMAND_BIT | TSL2591_REGISTER_CONTROL,
+                        int_time | gain);
+    if (rc) {
+        goto err;
+    }
+
+    /* Increment the gain change counter */
+    STATS_INC(g_tsl2591stats, gain_changed);
+
+    return 0;
+err:
+    return rc;
+}
+
+int
+tsl2591_get_gain(struct sensor_itf *itf, uint8_t *gain)
+{
+    int rc;
+    uint8_t reg;
+
+    rc = tsl2591_read8(itf, TSL2591_COMMAND_BIT | TSL2591_REGISTER_CONTROL,
+                       &reg);
+    if (rc) {
+        goto err;
+    }
+
+    *gain = reg & 0x30;
+
+    return 0;
+err:
+    return rc;
+}
+
+int
+tsl2591_get_data_r(struct sensor_itf *itf, uint16_t *broadband, uint16_t *ir)
+{
+    int rc;
+
+    #if MYNEWT_VAL(TSL2591_ITIME_DELAY)
+    /* Insert a delay of integration time to ensure valid sample */
+    os_time_delay((OS_TICKS_PER_SEC * g_tsl2591_itime_delay_ms) / 1000);
+    #endif
+
+    /* CHAN0 must be read before CHAN1 */
+    /* See: https://forums.adafruit.com/viewtopic.php?f=19&t=124176 */
+    *broadband = *ir = 0;
+    rc = tsl2591_read16(itf, TSL2591_COMMAND_BIT |
+                        TSL2591_REGISTER_CHAN0_LOW, broadband);
+    if (rc) {
+        goto err;
+    }
+    rc = tsl2591_read16(itf, TSL2591_COMMAND_BIT |
+                        TSL2591_REGISTER_CHAN1_LOW, ir);
+    if (rc) {
+        goto err;
+    }
+
+    /* Increment the polling counter */
+    STATS_INC(g_tsl2591stats, polled);
+
+    return 0;
+err:
+    return rc;
+}
+
+int
+tsl2591_get_data(struct sensor_itf *itf, uint16_t *broadband, uint16_t *ir)
+{
+    int rc;
+    uint8_t itime;
+    uint16_t igain;
+    uint16_t maxval;
+
+    #if !(MYNEWT_VAL(TSL2591_AUTO_GAIN))
+      return tsl2591_get_data_r(itf, broadband, ir);
+    #endif
+
+    /* Use auto-gain algorithm for better range at the expensive of */
+    /* unpredictable conversion times */
+
+    /* Get the integration time to determine max raw value */
+    rc = tsl2591_get_integration_time(itf, &itime);
+    if (rc) {
+        goto err;
+    }
+    /* Max value for 100ms = 37888, otherwise 65535 */
+    maxval = itime ? 65535 : 37888;
+
+    /* Set gain to 1x for the baseline conversion */
+    rc = tsl2591_set_gain(itf, TSL2591_LIGHT_GAIN_LOW);
+    if (rc) {
+        goto err;
+    }
+
+    /* Get the baseline conversion values */
+    /* Note: double-read required to empty cached values with prev gain */
+    rc = tsl2591_get_data_r(itf, broadband, ir);
+    rc = tsl2591_get_data_r(itf, broadband, ir);
+    if (rc) {
+        goto err;
+    }
+
+    /* Determine the ideal gain setting */
+    igain = maxval / (*broadband > *ir ? *broadband : *ir);
+
+    /* Find the closest gain <= igain */
+    if (igain < 25) {
+        /* Gain 1x, return the current sample */
+        return 0;
+    }
+    else if (igain < 428) {
```

Comment:
See https://github.com/apache/mynewt-core/blob/master/CODING_STANDARDS.md regarding whitespaces and braces.